core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -384,127 +387,100 @@ private boolean applyFailedShard(RoutingAllocation allocation, ShardRouting fail
-                        ShardRouting shardRouting = relocatingFromNode.next();
-                        if (shardRouting.equals(failedShard)) {
-                            dirty = true;
-                            if (addToIgnoreList) {
-                                // make sure we ignore this shard on the relevant node
-                                allocation.addIgnoreShardForNode(failedShard.shardId(), failedShard.currentNodeId());
-                            }
-                            relocatingFromNode.moveToUnassigned(unassignedInfo);
-                            break;
-                        }
-                    }
-                }
-                if (dirty) {
-                    logger.debug("failed shard {} found in routingNodes, failing it", failedShard);
-                    // next, we need to find the target initializing shard that is recovering from, and remove it...
-                    RoutingNodes.RoutingNodeIterator initializingNode = routingNodes.routingNodeIter(failedShard.relocatingNodeId());
-                    if (initializingNode != null) {
-                        while (initializingNode.hasNext()) {
-                            ShardRouting shardRouting = initializingNode.next();
-                            if (shardRouting.shardId().equals(failedShard.shardId()) && shardRouting.initializing()) {
-                                dirty = true;
-                                initializingNode.remove();
-                            }
-                        }
+        RoutingNodes.RoutingNodeIterator matchedNode = routingNodes.routingNodeIter(failedShard.currentNodeId());
+        if (matchedNode == null) {
+            logger.debug("{} ignoring shard failure, unknown node in {} ({})", failedShard.shardId(), failedShard, unassignedInfo.shortSummary());
+            return false;
+        }
+
+        boolean matchedShard = false;
+        while (matchedNode.hasNext()) {
+            ShardRouting routing = matchedNode.next();
+            if (routing.allocationId().getId().equals(failedShard.allocationId().getId())) {
+                matchedShard = true;
+                logger.debug("{} failed shard {} found in routingNodes, failing it ({})", failedShard.shardId(), failedShard, unassignedInfo.shortSummary());
+                break;
+            }
+        }
+
+        if (matchedShard == false) {
+            logger.debug("{} ignoring shard failure, unknown allocation id in {} ({})", failedShard.shardId(), failedShard, unassignedInfo.shortSummary());
+            return false;
+        }
+
+        // replace incoming copy to make sure we work on the latest copy
+        failedShard = matchedNode.current();
+
+        if (addToIgnoreList) {
+            // make sure we ignore this shard on the relevant node
+            allocation.addIgnoreShardForNode(failedShard.shardId(), failedShard.currentNodeId());
+        }
+
+        // handle relocation target shards
```

Comment:
can we add a better javadoc here? and that if we find it? Especially the early return statement at the end of this if, that one is confusing